kernel/byterun/coq_interp.c

```diff
@@ -837,31 +849,26 @@ value coq_interprete
       } /* fallthrough */
       Instruct(OFFSETCLOSURE) {
         print_instr("OFFSETCLOSURE");
-        accu = coq_env + *pc++ * sizeof(value); Next;
-      }
-      Instruct(PUSHOFFSETCLOSUREM2) {
-        print_instr("PUSHOFFSETCLOSUREM2");
-        *--sp = accu;
-      } /* fallthrough */
-      Instruct(OFFSETCLOSUREM2) {
-        print_instr("OFFSETCLOSUREM2");
-        accu = coq_env - 2 * sizeof(value); Next;
+        accu = coq_env - *pc++ * 3 * sizeof(value);
```

Comment:
Same remark as ENVACC, for OFFSETCLOSURE.
At leat *pc++ * 3 can be done at compile time.
It is true also for  "* sizeof(value)" but I think it is less important (due to x86 and ARM addressing mode).
As before, this is a very small optimisation.

Reply:
I agree it could be done at compile time, and I hesitated to do so. But I eventually decided against it. I think it is much more important that the compilation process is agnostic to the size of closures. It is better to leave this implementation detail to the interpreter. (Same thing for `ENVACC`, but there `+ 2` comes for free.)

Reply:
I think the most important is efficiency for the VM. Furthermore, the change is very small, since it is a trivial patch in
emitcode. I think I agree with you for ENVACC, the + 2 will come for free.

Reply:
> I think it is much more important that the compilation process is agnostic to the size of closures

Is there a way to have the interpreter expose this information in an explicit way? It would make it less risky (if the size of closures change in the interpreter, the corresponding constant should be updated, but the compiler does not need to be touched).

Reply:
Sur, the value can be exported, like for "accumulate"

Reply:
Let us be clear. This opcode is only called during a recursive call of a mutually recursive function with three siblings or more and this recursive call is not to one of the first two siblings. So, whichever optimization we perform there, I am not even sure we can actually notice an improvement for any of the existing Coq libraries.

As for the optimization itself, it removes the following assembly instruction:
```asm
leaq    (%rax,%rax,2), %rax
```

So, even if the patch to the compiler is only a few line long, I am still not convinced it is worth skipping one micro-operation (one fourth of a cycle) in a part of the code that is almost never executed.

Reply:
That said, looking at the assembly code, I have noticed that, since `*pc` is of type `int32_t`, there are a bunch of places where the generated assembly is not as efficient as possible on a 64-bit architecture (use of `cltq` and other sign-extension mechanisms). Indeed, the compiler is trying to preserve some kind of wraparound behavior there. But I will fix it in #12959 rather than here.

Reply:
Thank a lot, this code was developed before the generalisation of the 64-bit architecture.